Core/automation/lib/javascript/core/log.js

```diff
@@ -0,0 +1,199 @@
+                        _logger.error(this._getLogMessage(err));                        
+                    }
+                }},
+                
+                _getCallerDetails: { value: function _getCallerDetails (msg) {
+                    var matches = msg.stack.split('\n\tat ')[3].match(/(.+?) \((.+):(\d+?)\)/);
+                    msg.caller = matches[1];
+                    msg.fileName = matches[2]
+                    msg.lineNumber = matches[3]   
+                    return msg;
+                }},       
+                
+                _getLogMessage: { value: function _getLogMessage (msg, prefix, levelString) {                                        
+                    if ((typeof msg) !== "object") {
+                        msg = Error(msg);
+                        msg = this._getCallerDetails(msg);
+                    } else {
+                        msg.caller = msg.stack.split('\n\tat ')[1].split(' (')[0]
+                    }
+                    msg = this._legacyLoggerCorrection(msg);
+
+                    if (prefix === undefined) prefix = "log";                                                            
+                    
+                    if (prefix == "none") {
+                        return msg.message;
+                    }
+
+                    var level = "";
+                    var name = "";
+                    if (prefix != "log") {
+                        level = "[" + levelString + "] ";
+                        name = this._name === null ? "" : this._name;
+                    }
+
+                    if (prefix == "level") {
+                        return (level + msg);
+                    }
+
+                    var callerText;
+                    if (msg.caller.substr(0,1) == "<") {
+                        callerText = ", in " + msg.caller;
+                    } else {
+                        callerText = ", function " + msg.caller;
+                    }
+                    
+                    var message = msg.message == "" ? "" : "] " + msg.message;                    
+
+                    if (prefix == "short") {
+                        return (level + "[" + (name != "" ? name + ", " : "") + msg.fileName.split('/').pop() + ":" + msg.lineNumber + callerText + message);    
+                    }
+                    return (level + "[" + (name != "" ? name + ": " : "") + "source " + msg.fileName + ", line " + msg.lineNumber + callerText + message);
+                }},
+                
+                _legacyLoggerCorrection: { value: function _legacyLoggerCorrection (msg) {
+                    if (msg.fileName.search(/automation\/lib\/javascript\/core\/utils\.js$/) !== -1) {
+                        switch (msg.caller) {
+                            case "logError":
+                            case "logWarn":
+                            case "logInfo":
+                            case "logDebug":
+                            case "logTrace":
```

Comment:
This part was done specifically for support backwards compability support. If we get rid of logError, logWarn, ... in utils.log, this part goes away as well.

The new log functions are used like this:
```
load(__DIR__+'/log.js');
var log = Logger();
log.error("Message");
log.warn("Message");
log.info("Message");
log.debug("Message");
log.trace("Message");
```


utils.js does actually already provide this bridge between console.log and slf4j. I can move this part to log.js by simply instantiating a logger instance called console. `var console = Logger();`
Just need to add a few methods like `log()` and methods to change the logger name, log level and notification feature configuration, which currently can only be set at initialization.

But I also want to check, if we can so easily pretend to be console.log. Console.log also has a few tricks up it´s sleeve and I am not sure, if we can emulate all of that. But most I can remember is about color and font styles, which we just have to ignore / do not need to support.


I could convert log.js to use LogAction pretty easily instead of using slf4j directly, as LogAction does nothing else than wrap slf4j.
This  means loosing the level trace, which is no big issue. But it would make log.js slower. All the compilation of data about the log call position is very slow (a few milliseconds - forever for a logger). Therefore I check using slf4j's isErrorEnabled() and the likes, if it even makes sense to do all of this. Unfortunately LogAction does not expose these methods. (Actually, I am guilty of the same and should expose these in log.js.)

If we just want to look and behave like we are using LogAction under the hood, we could simply call our logger "org.eclipse.smarthome.model.script.jsr223.javascript" and get rid of the level trace.
Not sure, if LogAction adds some fancy formatting, which we would need to emulate.

I keep you posted on the console.log front.
Give me your thoughts on LogAction. Maybe we should try a pull request there to get the is<level>Enabled() methods.

Reply:
> But I also want to check, if we can so easily pretend to be console.log. Console.log also has a few tricks up it´s sleeve and I am not sure, if we can emulate all of that.

The big question is if there is any benefit to using this over LogAction :smile:. If concole.log has some cool stuff in it, then that may be what tips the scale. I think it is great to equalize the functionality between the JS and Jython libraries, but I really question if we need this (or the Jython) library at all. 

```
'use strict';
var OPENHAB_CONF = Java.type("java.lang.System").getenv("OPENHAB_CONF");
load(OPENHAB_CONF + '/automation/lib/javascript/core/actions.js');
LogAction.logWarn("Rules", "Test")
```
The Jython core.log provides [a helpful decorator](https://github.com/openhab-scripters/openhab-helper-libraries/blob/master/Core/automation/lib/python/core/log.py#L38) for getting a traceback, and it sends a notification too, but LogAction can do everything else. 

A big benefit to slf4j is the parameterized logging and only calculating the value to log if the log entry fits the logger. It's nice to have some performance heavy logging in place that doesn't slow things down when not in use. I haven't found a way to make use of this in Jython.

Reply:
As I mentioned elsewhere, I believe that there is benefit to having a console.log defined (so that existing JS code works), but what that forwards the actual log messages to I have no opinion on, and it could be either slf4j or LogAction with no impact to consumers.

Reply:
I had a look at console. Browser and node.js implementation.
The node.js is probably, the one we would like to emulate, as we are also a backend implementation.
But looking at the node.js documetation at https://nodejs.org/api/console.html, I would say console is a different, much bigger beast. Already the message formatting syntax is completely different and then there are quite some methods to be implemented.

Therefore I do not think that what I built should pretend to be console. At least the formatting syntax would need to be supported, then it would be just about features not available, but not differences in functionality.

We could make quick progress when we canibalize the node.js implementation, but I am not sure, if I am up for that right now. Maybe.